src/methods.rs

```diff
@@ -92,6 +98,28 @@ impl LateLintPass for MethodsPass {
                     }
                 }
             }
+            // check Option.map(_).unwrap_or(_)
+            else if name.node.as_str() == "unwrap_or" {
+                if let ExprMethodCall(ref inner_name, _, ref inner_args) = args[0].node {
+                    if inner_name.node.as_str() == "map"
+                            && match_type(cx, cx.tcx.expr_ty(&inner_args[0]), &OPTION_PATH) {
+                        span_lint(cx, OPTION_MAP_UNWRAP_OR, expr.span,
+                                  "called `map(f).unwrap_or(a)` on an Option value. This can be \
+                                   done more directly by calling `map_or(a, f)` instead");
+                    }
+                }
+            }
+            // check Option.map(_).unwrap_or_else(_)
+            else if name.node.as_str() == "unwrap_or_else" {
+                if let ExprMethodCall(ref inner_name, _, ref inner_args) = args[0].node {
+                    if inner_name.node.as_str() == "map"
+                            && match_type(cx, cx.tcx.expr_ty(&inner_args[0]), &OPTION_PATH) {
+                        span_lint(cx, OPTION_MAP_UNWRAP_OR_ELSE, expr.span,
+                                  "called `map(f).unwrap_or_else(g)` on an Option value. This can \
+                                   be done more directly by calling `map_or_else(g, f)` instead");
+                    }
```

Comment:
I thought about doing something like that, but my worry was that if the arguments to `map_or`/`map_or_else` were verbose (e.g. the code I changed in src/mut_mut.rs), then it could lead to a very cluttered-looking lint message. Does snippet shorten things in long cases like that?